### PR TITLE
Modify Expand to support -1 dim value [Release 1.7]

### PR DESCRIPTION
**Pull request by neginraoof**
SImilar to Pytorch expand op, support shape input with -1 dim value, in which case the actual dimension value will remain unchanged.

### CODE REVIEW

**Comment by linkerzhang** [comment]
@ebarsoum @postrational "-1" has different meanings across ONNX ops. It may be good to have a consistent definition across all of them to avoid confusion. See issue #2507

**Comment by spandantiwari** [comment]
@neginraoof  - would it be OK to use '0' instead of '-1' for keeping the dimension unchanged. I think there are other ops that are use 0 for that semantics. We should try to be consistent. 

@linkerzhang - Issue https://github.com/onnx/onnx/issues/2507 has been open for a while, and I see multiple options open for that. Do you know if any decision has been made on that? 
If yes, then we can follow the same semantics here.

**Comment by spandantiwari** [comment]
cc: @houseroad

**Comment by gramalingam** [comment]
My understanding from issue 2507 is that using 0 is problematic, and that -1 is better.

File: docs/Operators.md [hunk 1]
```diff
@@ -5146,6 +5151,40 @@ expect(node, inputs=[data, new_shape], outputs=[expanded],
...
+<summary>dim_negative</summary>
+
+```python
+node = onnx.helper.make_node(
+    'Expand',
+    inputs=['data', 'new_shape'],
+    outputs=['expanded'],
+)
+shape = [3, 3, 1]
+new_shape = [-1, 3, 6]
```
Comment (gramalingam) [inline docs/Operators.md]: Why can't we use new_shape = [1, 3, 6] instead?

Comment (neginraoof) [reply docs/Operators.md]: This test is for the case when we have -1 in expand shape input.

**Comment by neginraoof** [comment]
Just a note:
I recently updated constant of shape to create multidimensional tensors with dim value 0.
https://github.com/onnx/onnx/pull/2567

File: docs/Operators.md (hunk 1 above)
Comment (gramalingam) [reply docs/Operators.md]: I was using this to illustrate my previous question: why do we need this extension of the op, if we can use "1" instead? Am I misunderstanding something?

Comment (neginraoof) [reply docs/Operators.md]: We do not always know the dim value we want to expand to.
Let's say you have a tensor of size (?, ?, 1) and you wan to expand to (?, ?, 5) where ? could be unknown.

Comment (gramalingam) [reply docs/Operators.md]: Then, you just specify the new-shape to be (1, 1, 5) ... if you apply this to a tensor of size (M,N,1), won't you end up with the shape (M, N, 5)?

Comment (neginraoof) [reply docs/Operators.md]: No,  you cannot expand (M, N, 1) tensor to (1, 1, K)

**Comment by neginraoof** [comment]
@linkerzhang 
Thanks for pointing the PR. I believe -1 is a good choice, specially if we decide to allow zero.

File: docs/Operators.md (hunk 1 above)
Comment (gramalingam) [reply docs/Operators.md]: Bidirectional broadcast between the shapes (1, 1, K) and (M, N, 1) will produce the shape (M, N, K). And I believe Expand (the existing version) does bidirectional-broadcast, as far as I can see. So, what am I missing?

Comment (gramalingam) [reply docs/Operators.md]: My understanding from the existing documentation is that ONNX Expand differs from numpy's broadcast_to in that ONNX uses bidirectional broadcast, unlike numpy.

Comment (neginraoof) [reply docs/Operators.md]: This is more of an export issue, since for exporting PT or TF models that allow -1, we need to change the shape input tensor.

Comment (gramalingam) [reply docs/Operators.md]: I see. So, the semantics of "-1" will be the same as that of "1" (in ONNX)? (Not sure if PT or TF have the exact same semantics for 1 ... it does look odd to have two ways of saying the same thing.)

**Comment by neginraoof** [comment]
Closing this for now. As suggested by @gramalingam, we can have a workaround in export.

**Pull request closed without merging**